Changes behavior of builds to handle naming collisions

oatmealraisin opened the pull request:
When getting the next name for a build, check to see if the name already exists. If so, re-increment until success.

This is a WIP to check tests for failure.

[test][testextended][extended:core(builds)]

liggitt commented on pkg/build/generator/generator.go:
```diff
@@ -432,6 +432,12 @@ func (g *BuildGenerator) generateBuildFromConfig(ctx kapi.Context, bc *buildapi.
 	}
 
 	build.Name = getNextBuildName(bc)
+	if _, err := g.Client.GetBuild(ctx, build.Name); err == nil {
```
I think the resulting conflict is intentional to make sure that we don't race and create two builds accidentally
@csrwng

oatmealraisin commented:
@csrwng I've moved getNextBuildName to use namer.GetName, with the cap being at 63 characters. Also, I added a warning to notify the user of the naming collision. ptal?

csrwng commented:
@oatmealraisin please use kvalidation.DNS1123SubdomainMaxLength instead of 63

csrwng commented on pkg/build/generator/generator.go:
```diff
@@ -432,6 +432,14 @@ func (g *BuildGenerator) generateBuildFromConfig(ctx kapi.Context, bc *buildapi.
 	}
 
 	build.Name = getNextBuildName(bc)
+	// Retry getting a non-conflicting name 10 times
```
Can you add an explanation here of what happens when trying the next 10 numbers fails?

csrwng commented on pkg/build/generator/generator.go:
```diff
@@ -432,6 +432,14 @@ func (g *BuildGenerator) generateBuildFromConfig(ctx kapi.Context, bc *buildapi.
+	// Retry getting a non-conflicting name 10 times
+	if _, err := g.Client.GetBuild(ctx, build.Name); err == nil {
+		for i := 0; i < 10 && err == nil; i++ {
+			glog.V(5).Infof("A naming collision occurred with %s, retrying", build.Name)
```
Please add the namespace to the log, otherwise we won't know what build we're talking about

csrwng commented on pkg/build/generator/generator.go:
```diff
@@ -691,7 +699,7 @@ func resolveError(kind string, namespace string, name string, err error) error {
 func getNextBuildName(buildConfig *buildapi.BuildConfig) string {
 	buildConfig.Status.LastVersion++
-	return fmt.Sprintf("%s-%d", buildConfig.Name, buildConfig.Status.LastVersion)
+	return namer.GetName(buildConfig.Name, strconv.FormatInt(buildConfig.Status.LastVersion, 10), kvalidation.DNS1123SubdomainMaxLength)
```
Why not strconv.ItoA ? seems less verbose

csrwng replied:
ahh ok. cool

oatmealraisin commented:
@csrwng Ready for another look, I added the comment and namespace mention.

csrwng commented on pkg/build/generator/generator.go:
```diff
@@ -432,6 +432,17 @@ func (g *BuildGenerator) generateBuildFromConfig(ctx kapi.Context, bc *buildapi.
 
 	build.Name = getNextBuildName(bc)
+	// Retry getting a non-conflicting name 10 times. If, after 10 tries, there
+	// is still a naming conflict, the process will continue until we try and
```
shouldn't it say that "if after 10 tries there is still a naming conflict, we will fail, but the BuildConfig will be saved with the +10 increase" ?

csrwng replied:
nm, I get what you're saying

csrwng replied:
"the process will continue" in my mind meant "keep trying more names" ... maybe there's better wording for that.

oatmealraisin replied:
I'm trying not to imply that we should return an error there. I've updated the comment to be more clear about saving the LastVersion, ptal?

csrwng replied:
Latest version LGTM

oatmealraisin commented:
@csrwng I added/fixed tests, ptal?

liggitt commented on pkg/build/generator/generator.go:
```diff
@@ -432,6 +432,17 @@ func (g *BuildGenerator) generateBuildFromConfig(ctx kapi.Context, bc *buildapi.
+	// get a unique name. If that fails, keep the +10 version increase without
+	// reporting an error. When the Build is created, it will fail with a naming
+	// error
+	if _, err := g.Client.GetBuild(ctx, build.Name); err == nil {
```
Is this the right level to be iterating until we find a free name? Are all callers intending to generate a new build even if one already exists?

csrwng commented:
@oatmealraisin you shouldn't do a separate get to figure out if there's a collision. You can check the error when the create fails.

oatmealraisin commented:
@csrwng Do we want to keep increasing the LastVersion on failure?

csrwng commented:
Yes

csrwng commented:
@oatmealraisin please use IsAlreadyExists to check the error:
https://github.com/openshift/origin/blob/master/vendor/k8s.io/kubernetes/pkg/api/errors/errors.go#L370

oatmealraisin commented:
@csrwng I've changed it to telling them to change the LastVersion of the BuildConfig, are there any better ways of fixing this?

I could add a mention of `oc edit`, but I'm concerned about the length of text we're printing.

csrwng commented:
@oatmealraisin because we are incrementing the status.lastVersion, the only action you should need to take is to start the build again. Also, can you factor out the error message into a function and call it from each of the 3 places?

oatmealraisin commented:
@csrwng I've changed the message and moved the transform to a new function at the bottom of the file, ptal?

csrwng commented:
LGTM for post 1.4

oatmealraisin commented:
@csrwng merge?

csrwng commented:
[merge]
